src/goose/toolkit/web_browser.py

```diff
@@ -0,0 +1,441 @@
+            shadow_element = shadow.find_element_by_xpath(f"//*[contains(text(), '{text}')]")
+            if shadow_element:
+                self.notifier.notify(f"Element found in shadow DOM with text: {text}")
+                return shadow_element.get_attribute("outerHTML")
+        except Exception as e:
+            self.notifier.notify(f"Error searching in shadow DOM: {str(e)}")
+
+        self.notifier.notify(f"Element not found with text: {text} in either DOMs")
+        return None
+
+    @tool
+    def find_elements_of_type(self, tag_type: str, filename: str) -> list[str]:
+        """Find all elements of a specific tag type using BeautifulSoup on HTML content stored in a file.
+
+        Args:
+            tag_type (str): The HTML tag type to search for.
+            filename (str): The name of the file containing the HTML content.
+        """
+        elements_as_strings = []
+        try:
+            with open(filename, "r", encoding="utf-8") as file:
+                soup = BeautifulSoup(file, "html.parser")
+                elements = soup.find_all(tag_type)
+                elements_as_strings = [str(element) for element in elements]
+                self.notifier.notify(f"Found {len(elements_as_strings)} elements of type: {tag_type}")
+        except FileNotFoundError:
+            self.notifier.notify(f"File not found: {filename}")
+        return elements_as_strings
+
+    def __del__(self) -> None:
+        # Remove the entire session directory
+        if os.path.exists(self.session_dir):
+            try:
+                shutil.rmtree(self.session_dir)
+                self.notifier.notify(f"Removed browsing session directory: {self.session_dir}")
+            except OSError as e:
+                self.notifier.notify(f"Error removing session directory: {str(e)}")
+
+        if self.driver:
+            self.driver.quit()
+
+
+def get_default_browser_windows() -> str:
+    try:
+        with winreg.OpenKey(
+            winreg.HKEY_CURRENT_USER, r"Software\Microsoft\Windows\Shell\Associations\UrlAssociations\http\UserChoice"
+        ) as key:
+            prog_id, _ = winreg.QueryValueEx(key, "ProgId")
+
+        with winreg.OpenKey(winreg.HKEY_CLASSES_ROOT, f"{prog_id}\\shell\\open\\command") as cmd_key:
+            command, _ = winreg.QueryValueEx(cmd_key, None)
+
+        if command.startswith('"'):
+            executable = command.split('"')[1]
+        else:
+            executable = command.split(" ")[0]
+
+        return os.path.basename(executable)
+
+    except Exception as e:
+        print(f"Error retrieving default browser on Windows: {e}")
+        return None
+
+
+def get_default_browser_macos() -> str:
+    try:
+        try:
+            result = subprocess.run(
+                "defaults read"
+                + " ~/Library/Preferences/com.apple.LaunchServices/com.apple.launchservices.secure"
+                + " LSHandlers -array | grep -B 1 -A 1 http | grep LSHandlerRoleAll | awk -F= '{print $2}'"
```

Comment:
instead of this `grep`/`awk` stuff can we use the python `pslistlib` lib for this?

easier to maintain

```python3
for h in handlers:
  scheme = h.get('LSHandlerURLScheme')
  if scheme and 'http' in scheme:
      return x.get('LSHandlerRoleAll')
```

Reply:
on it